### PR TITLE
[+] Add insertEntry usecase

### PR DESCRIPTION
**Pull request by Yunshan-522**

### CODE REVIEW

**Review by Yuhan-Ut — COMMENTED** [review]
I see you add another method in this interface, but lots of other classes implement this interface, since we can not make a abstract viewmode for now, I think it's better to create your own interface for your viewmodel (temporarily solution ), so this will not cause errors in other classes in other use-cases (as what is warning in the auto-check)

File: src/main/java/billgates/interface_adapters/Updatable.java [hunk 1]
```diff
@@ -1,12 +1,15 @@
...
 import billgates.usecases.bill_update.BillUpdateViewModel;
+import billgates.usecases.insert_entry.InsertEntryViewModel;
 
-// Interace Adapter Layer
+// Interface Adapter Layer
 public interface Updatable {
 
     void update(BillUpdateViewModel viewModel);
 
+    void insert(InsertEntryViewModel viewModel);
```
[the remark this answers is not in the document]
Comment (Yunshan-522) [reply src/main/java/billgates/interface_adapters/Updatable.java:11]: Okay! I will change it!

**Review by Ellen141319 — COMMENTED** [review]
Great job! I think it is a good way to reduce the dependency.

File: src/main/java/billgates/usecases/insert_entry/InsertEntryPresenter.java [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+public class InsertEntryPresenter implements InsertEntryOutputPort{
+
+    private final Updatable_insert view;
+
+    public InsertEntryPresenter(Updatable_insert view) {
+        this.view = view;
+    }
+
+    @Override
+    public void insertEntry(InsertEntryResponseModel model){
```
Comment (Lei-Tin) [inline src/main/java/billgates/usecases/insert_entry/InsertEntryPresenter.java:17]: Similar problem in this pull request as the last one that I have reviewed [here](https://github.com/CSC207-2022F-UofT/course-project-bill-gates/pull/35).

To my knowledge, we do not need to implement a `ViewModel` or `output port` that is used to update the display or the presenter. The purpose of this `InsertEntryUseCase` is simply to insert the entry that we have obtained from the `InputPort` into the database by calling methods that is used to interact with it.

File: src/main/java/billgates/usecases/insert_entry/InsertEntryUseCase.java [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+    public InsertEntryUseCase(InsertEntryOutputPort presenter, DatabaseGateway gateway, BillUpdateUseCase update){
+        this.presenter = presenter;
+        this.gateway = gateway;
+        this.update = update;
+    }
+
+    @Override
+    public void insertEntry(QueryEntryData entry) {
+        gateway.insertEntry(entry.getId(), entry);
+        update.updateBill();
```
Comment (Lei-Tin) [inline src/main/java/billgates/usecases/insert_entry/InsertEntryUseCase.java:26]: Similar problem as mentioned above, but what I'm stating might not be the final resolution yet, we have to reach a decision with Scott

File: src/main/java/billgates/usecases/insert_entry/InsertEntryUseCase.java [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+        List<Entry> result = this.gateway.getBillData(User.getInstance().getCurrentBillID()).getEntries()
+                .stream().map(d -> d.toEntryBuilder().build()).toList();
+        System.out.println(result); // TODO remove this
```
Comment (Lei-Tin) [inline src/main/java/billgates/usecases/insert_entry/InsertEntryUseCase.java:29]: Also, why isn't this removed XD

**Review by Lei-Tin — COMMENTED** [review]
Everything here is looking pretty good! 👍 

Just that, maybe we do not need the presenter for this specific use case, as we are calling this use case only for the purpose of inserting an entry into the database, the part of interacting with the GUI will be done by the `BillUpdateUseCase`

Something extra nice I noticed about this PR is that, the conventions are being held pretty nicely! Good job!

**Comment by ScottCTD** [comment]
The only problem is that we actually don't need a presenter, view interface, and view model for this use case.

As illustrated in https://github.com/CSC207-2022F-UofT/course-project-bill-gates/discussions/37#discussioncomment-4185696, we can call the `UpdateBillUseCase` after your insertEntry use case finishes its job.

**Comment by ScottCTD** [comment]
> The only problem is that we actually don't need a presenter, view interface, and view model for this use case.
> 
> As illustrated in [#37 (comment)](https://github.com/CSC207-2022F-UofT/course-project-bill-gates/discussions/37#discussioncomment-4185696), we can call the `UpdateBillUseCase` after your insertEntry use case finishes its job.

Therefore, an easy way to fix this is simply to remove the view interface, view model, and the presenter.

**Comment by Yuhan-Ut** [comment]
After checking my use-case implements and discussing with group members, now I have deeper understanding about usecases, so please let me to redo my review about this pull request.

First of all, everything is looking good here and I think this use case will work as we expected! 

But as a use case that change recorded information in the database, I may assume there is no need for output port, viewmodel or presenter, we can left the job to be done by the update_bill use case, so duplicate code is avoided.

Sorry for the misleading I may made, and hope you will keep on going with this excellent job.

ps. some quick tips for fixing the presenter issue, remove view interface, view model, and the presenter and the lines I will mark in the usecase.java file

File: src/main/java/billgates/usecases/insert_entry/InsertEntryUseCase.java [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+        this.presenter = presenter;
+        this.gateway = gateway;
+        this.update = update;
+    }
+
+    @Override
+    public void insertEntry(QueryEntryData entry) {
+        gateway.insertEntry(entry.getId(), entry);
+        update.updateBill();
+        List<Entry> result = this.gateway.getBillData(User.getInstance().getCurrentBillID()).getEntries()
```
Comment (Yuhan-Ut) [inline src/main/java/billgates/usecases/insert_entry/InsertEntryUseCase.java:27]: you may want to remove every line below and including this line in this file and view-model, presenter, and output port to remove the presenter function which is not necessary for this use-case.
once again, you did very good job!

**Pull request closed without merging**